Use Nimble options. Validation in prepare_for_start.

koozdra opened the pull request:
I moved adding the configuration to persistent term to the prepare for start callback of the producer (should this still be in the client?).

I wasn't able to use Nimble options to validate some of the integer max values. I added TODOs in the tests where the validation is still needed. The validation for the allowed values of the `attribute_names` is missing.

Should I add custom validation in this repo or extend Nimble opts to be able to have a :list type with a check on every member and the ability to have validation for max on integer types?

Another issue here is that Broadway in mix.exs is point to broadway master. Not sure which tag or version I should use.

msaraiva commented:
Hi @koozdra!

Thanks for the PR.
 
> I moved adding the configuration to persistent term to the prepare for start callback of the producer (should this still be in the client?)

I believe not. However, I'm not sure this should be in `prepare_for_start/2`. I think it could be on `init/1` as it doesn't read or update anything from the general topology options. WDYT?

> I wasn't able to use Nimble options to validate some of the integer max values...
> Should I add custom validation in this repo or extend Nimble opts to be able to have a :list type with a check on every member and the ability to have validation for max on integer types?

I think our best option here is to create a custom type. You can see a few examples here: https://github.com/dashbitco/nimble_options/blob/master/test/nimble_options_test.exs#L547-L596

> The validation for the allowed values of the `attribute_names` is missing.

This one could use the built-in type `{:one_of, choices}`.

josevalim commented:
@msaraiva I was the one who suggested doing it in `prepare_for_start` so we validate it only once, instead of once per producer. But perhaps this is a silly optimization and doing it on `init` is clearer? What are your thoughts?

msaraiva commented:
@josevalim I think validation on `init` is something we already expect and, as you said, the optimization for this particular case will be probably irrelevant. As for `prepare_for_start`, I think we don't expect it to fail, do we? Otherwise, it would be better to make that explicit and maybe change the return value to `{:ok, ...} | {:error, ...}`. Not sure if it's worth it though.

josevalim commented:
We are raising on prepare_for_start, so we don’t need an error return. I want to point out that raising from prepare_for_start has one benefit: the error message is clearer because it is less nested inside processes. Given that the error is clearer and we run the validation less times, I think it is worth giving it a try!

msaraiva commented:
> the error message is clearer because it is less nested inside processes

That's a very good point. Let's go for it 👍 

Maybe we should update the [docs](https://hexdocs.pm/broadway/Broadway.Producer.html#c:prepare_for_start/2) to make that explicitly too:

> The goal of this task is to manipulate the general topology options, **validate them**, if necessary at all, and introduce any new child specs that will be started before the ProducerSupervisor in Broadwday's supervision tree.

koozdra commented:
@msaraiva Thanks for the feedback

Fixed all tests with custom validators.

> This one could use the built-in type {:one_of, choices}.

We are checking the members of a list to be a `:one_of` choice, is that possible with Nimble options? In the mean time I created a `type_list_limited_member` custom validator that will do the check.

> Why does it have to point to master? I'm probably missing something.

Should I point to the https://github.com/dashbitco/broadway/releases/tag/v0.6.2 tag for the Broadway dependancy? Right now I'm pointing to master to get the latest Broadway code to make the necessary changes.

msaraiva commented:
> We are checking the members of a list to be a `:one_of` choice, is that possible with Nimble options? In the mean time I created a `type_list_limited_member` custom validator that will do the check.

Actually, it's not possible yet. We do have a [PR](https://github.com/dashbitco/nimble_options/pull/35) for it though. It seems to me that the `{:list, type}` can be quite useful so my vote is to add it. @josevalim WDYT? I believe the PR is ready to be merged. 

> Should I point to the https://github.com/dashbitco/broadway/releases/tag/v0.6.2 tag for the Broadway dependancy? Right now I'm pointing to master to get the latest Broadway code to make the necessary changes.

Are you using any feature that was added after `v0.6.2`? If not, yes, just stick with the latest version released on hex. Otherwise, we'll need to release a new version of Broadway.

On last thing we need to do is to move all [options'  documentation](https://github.com/dashbitco/broadway_sqs/blob/fb59c66a1400b2f387e698ae431af7cc7763d453/lib/broadway_sqs/producer.ex#L15-L60) to `BroadwaySQS.Options.definition/0` and use `NimbleOptions.Docs.generate/1` to generate it as it was done for Broadway [here](https://github.com/dashbitco/broadway/blob/f1cc5e3a215a5dfc93d7e37de6f1be87adb210f1/lib/broadway.ex#L754).

koozdra commented:
@msaraiva 

Switched to Broadway 0.6.2. I missed the requirement that `attribute_names` and `message_attribute_names` is allowed to be `:all` as well as a list. I added tests for those situations. I moved the documentation for the options into the options module and used that to generate the docs for the producer module.

Taking a look at the ci failures..

msaraiva commented on lib/broadway_sqs/producer.ex:
```diff
@@ -12,52 +12,7 @@ defmodule BroadwaySQS.Producer do
+  #{NimbleOptions.Docs.generate(BroadwaySQS.Options.definition())}
 
   ## Producer Options
 
```
I think we should move the docs from the following two items to the options' definition and remove the "Options for BroadwaySQS.ExAwsClient" section. This way we can have a single "Options" section like:

```
  ## Options

  Aside from `:receive_interval` and `:sqs_client` which are generic and apply to all
  producers (regardless of the client implementation), all other options are specific to
  the `BroadwaySQS.ExAwsClient`, which is the default client.
 
  #{NimbleOptions.Docs.generate(BroadwaySQS.Options.definition())}

msaraiva commented on lib/broadway_sqs/options.ex:
```diff
@@ -0,0 +1,202 @@
+        "Acknowledgments" section below for all the possible values. Defaults to `:ack`.
+        """
+      ],
+      on_failure: [
```
We're missing the doc of this one :)

msaraiva reviewed (commented):
Hey @koozdra!

Fantastic job! I left just a couple of comments regarding the docs.

koozdra commented:
@msaraiva 
👍 documentation fixes

msaraiva commented on lib/broadway_sqs/options.ex:
```diff
@@ -0,0 +1,217 @@
+  end
+
+  def type_non_empty_string("", [{:name, name}]) do
+    {:error, "expected :#{name} to be an string, got: \"\""}
```
```suggestion
    {:error, "expected :#{name} to be a non-empty string, got: \"\""}
```

msaraiva commented on lib/broadway_sqs/options.ex:
```diff
@@ -0,0 +1,217 @@
+  end
+
+  def type_non_empty_string(value, [{:name, name}]) do
+    {:error, "expected :#{name} to be an string, got: #{inspect(value)}"}
```
```suggestion
    {:error, "expected :#{name} to be a non-empty string, got: #{inspect(value)}"}
```

msaraiva commented on test/broadway_sqs/producer_test.exs:
```diff
@@ -60,14 +68,451 @@ defmodule BroadwaySQS.BroadwaySQS.ProducerTest do
+
+  describe "prepare_for_start/2 validation" do
+    test "when the queue url is not present" do
+      assert_raise(
+        ArgumentError,
+        "invalid configuration given to SQSBroadway.prepare_for_start/2, required option :queue_url not found, received options: []",
+        fn ->
+          prepare_for_start_module_opts([])
+        end
+      )
```
Since the formatter won't touch those large one-line strings, maybe could do:

```suggestion
      message = """
      invalid configuration given to SQSBroadway.prepare_for_start/2, \
      required option :queue_url not found, received options: []\
      """
      
      assert_raise(ArgumentError, message, fn ->
        prepare_for_start_module_opts([])
      end)
```

msaraiva commented on test/broadway_sqs/producer_test.exs:
```diff
@@ -60,14 +68,451 @@ defmodule BroadwaySQS.BroadwaySQS.ProducerTest do
+    test "when the queue url is nil" do
+      assert_raise(
+        ArgumentError,
+        "invalid configuration given to SQSBroadway.prepare_for_start/2, expected :queue_url to be an string, got: nil",
```
From this point on, I think you we could do just:

```suggestion
        ~r/expected :queue_url to be a non-empty string, got: nil/,
```

and then in case the string is much larger, just create a variable and use `~r"""`. Similiar to the previous suggestion. WDYT?

koozdra replied:
Love the ideas and changes implemented.

msaraiva reviewed (commented):
Looks great! Just a couple of nitpicks and we're good to go.

Sorry for not mentioning them before :)

msaraiva commented:
@koozdra thanks! 🎉

leandrocp commented:
> Actually, it's not possible yet. We do have a [PR](https://github.com/dashbitco/nimble_options/pull/35) for it though. It seems to me that the `{:list, type}` can be quite useful so my vote is to add it.

@msaraiva that PR was mostly done but I've closed it because looked like it was not the moment to add that validation. But I can reopen it whenever necessary, just ping me.
